### ISSUE TITLE
Reentering the campaign results in two logs scheduled for the same contact

### ISSUE THREAD

**Issue by giomasce** [mautic/mautic#9797]
## Bug Description
Under certain conditions, leads reentering a campaign which they are already going through end up having two logs scheduled at the same time. This eventually leads to Mautic trying to insert duplicated logs, failing and stalling leads in the campaign (see for example https://github.com/mautic/mautic/issues/7953#issuecomment-794430819).

| Q   | A
| --- | ---
| Mautic version | From `features` branch
| PHP version | 7.4.15 from Debian testing in test virtual machine
| Browser | Firefox 86.0.1

### Steps to reproduce
1. Create a campaign with the following properties.
    * Leads are taken from a campaign form
    * Leads are allowed to restart the campaign
    * First event, executed immediately: send an email
    * Second event, executed after one day, on Thursday, at 07:00: send an email
    * Third event, executed after one day: send an email
    * Fourth event, executed after one day: send an email
2. Fill the form with a certain email, on Wednesday (17/03/2021 in my case); trigger the campaign, which results in the first email being sent
3. Advance time to Thursday (18/03/2021; Mautic is installed inside a virtual machine, so I can go to the future), past 7 UTC (corresponding to 8 CET in my timezone); trigger the campaign, which results in the second email being sent
4. Fill the form again with the same email; trigger the campaign, which results in the first email being sent again
5. Notice that now two emails are scheduled for the contact: the second email (Thursday) is scheduled for 25/03/2021 and rotation 1, and the third one (Friday) is scheduled for 19/03/2021 and rotation 2. This seems wrong: I would have expected the Friday scheduled log to be removed or unscheduled.
 
### Log  errors

[mautic_prod-2021-03-17.txt](https://github.com/mautic/mautic/files/6157970/mautic_prod-2021-03-17.txt)
[mautic_prod-2021-03-18.txt](https://github.com/mautic/mautic/files/6157971/mautic_prod-2021-03-18.txt)

**Comment by kuzmany** [comment]
Can you try these two fixes?

https://github.com/mautic/mautic/pull/9639 
https://github.com/mautic/mautic/pull/9486

**Comment by giomasce** [comment]
Hi @kuzmany, thanks for your reply. Would you suggest to try them together, or one a time?

**Comment by kuzmany** [comment]
You can try one a time.

**Comment by giomasce** [comment]
Neither of the two fixed my problem, unfortunately.

**Comment by giomasce** [comment]
I prepared a more detailed transcript of a session that shows the bug, including the integrity exception caused by the database. The campaign is set up as detailed in the original bug description.

### 22/03/2021: subscribed a contact to the campaign
* First email is sent
* Database dump
![snap1](https://user-images.githubusercontent.com/101675/112033057-5299a900-8b3d-11eb-8188-16c38b5e48cb.png)
* Log:
```
[2021-03-22 17:14:47] mautic.DEBUG: CAMPAIGN: Triggering kickoff events  
[2021-03-22 17:14:47] mautic.DEBUG: CAMPAIGN: Processing the following events: 1  
[2021-03-22 17:14:47] mautic.DEBUG: CAMPAIGN: No events to process  
[2021-03-22 17:14:47] mautic.DEBUG: CAMPAIGN: Triggering scheduled events  
[2021-03-22 17:14:47] mautic.DEBUG: CAMPAIGN: 0 events scheduled to execute.  
[2021-03-22 17:14:47] mautic.DEBUG: CAMPAIGN: No events to process  
[2021-03-22 17:14:47] mautic.DEBUG: CAMPAIGN: Triggering inaction events  
[2021-03-22 17:14:47] mautic.DEBUG: CAMPAIGN: No events to process  
[2021-03-22 17:15:49] mautic.DEBUG: CONTACT: In a Mautic user session  
[2021-03-22 17:15:49] mautic.DEBUG: CONTACT: In a Mautic user session  
[2021-03-22 17:15:49] mautic.DEBUG: FORM: Not in kiosk mode so using current contact ID #  
[2021-03-22 17:15:49] mautic.DEBUG: FORM: Unique fields submitted include test@mascellani.eu  
[2021-03-22 17:15:49] mautic.DEBUG: FORM: Potential conflicts 0 = email  
[2021-03-22 17:15:49] mautic.DEBUG: FORM: Current unique contact fields email = test@mascellani.eu  
[2021-03-22 17:15:49] mautic.DEBUG: FORM: Submitted unique contact fields email = test@mascellani.eu  
[2021-03-22 17:15:49] mautic.DEBUG: FORM: Associating 127.0.0.1 to contact  
[2021-03-22 17:15:49] mautic.DEBUG: CONTACT: 1 set as current lead.  
[2021-03-22 17:15:49] mautic.DEBUG: LEAD: 1 set as system lead.  
[2021-03-22 17:15:49] mautic.DEBUG: CAMPAIGN: Contact ID 1 was added to campaign ID 1 as a new member.  
[2021-03-22 17:15:49] mautic.DEBUG: CAMPAIGN: Campaign triggered for event type form.submit(form / 1)  
[2021-03-22 17:15:49] mautic.DEBUG: CONTACT: System lead is being used  
[2021-03-22 17:15:49] mautic.DEBUG: CAMPAIGN: Current contact ID# 1  
[2021-03-22 17:15:49] mautic.DEBUG: CAMPAIGN: Contact does not have any applicable form.submit associations.  
[2021-03-22 17:16:46] mautic.DEBUG: CAMPAIGN: Triggering kickoff events  
[2021-03-22 17:16:46] mautic.DEBUG: CAMPAIGN: Processing the following events: 1  
[2021-03-22 17:16:46] mautic.DEBUG: CAMPAIGN: Processing the following contacts: 1  
[2021-03-22 17:16:46] mautic.DEBUG: CAMPAIGN: (1) Executing immediately  
[2021-03-22 17:16:46] mautic.DEBUG: CAMPAIGN: Event ID# 1 to be executed on 2021-03-22 17:16:46 UTC compared to 2021-03-22 17:16:46 UTC  
[2021-03-22 17:16:46] mautic.DEBUG: CAMPAIGN: (1) Executing immediately  
[2021-03-22 17:16:46] mautic.DEBUG: CAMPAIGN: Event ID# 1 to be executed on 2021-03-22 17:16:46 UTC  
[2021-03-22 17:16:46] mautic.DEBUG: CAMPAIGN: Executing email.send ID 1  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: Evaluating 0 conditions for action ID 1  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: Executing 1 actions under action ID 1  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: (2) Adding interval of 1d to 2021-03-22 17:16:46 UTC  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: (2) 2021-03-23 17:16:46 UTC is later than 2021-03-22 17:16:46 UTC and thus returning 2021-03-23 17:16:46 UTC  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: Event ID# 2 to be executed on 2021-03-23 17:16:46 UTC  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: Comparing calculated executed time for event ID 2 and contact ID 1 with 2021-03-22 17:16:47 UTC  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: Scheduling event ID 2 for contact ID 1 based on hour of 07:00 UTC  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: Scheduling event ID 2 for contact ID 1 based on DOW restrictions of 4  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: Action ID# 2 for contact ID# 1 has timing that is not appropriate and thus scheduled for 2021-03-25 07:03:00 UTC  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: Fetching the next batch of kickoff contacts starting with contact ID 2  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: No more contacts to process  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: Triggering scheduled events  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: 0 events scheduled to execute.  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: No events to process  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: Triggering inaction events  
[2021-03-22 17:16:47] mautic.DEBUG: CAMPAIGN: No events to process  
```

### 25/03/2021 (Thursday)
* Second email is sent
* Database dump:
![snap2](https://user-images.githubusercontent.com/101675/112033234-81b01a80-8b3d-11eb-8f82-56c59ab8106d.png)
* Log:
```
[2021-03-25 07:00:11] mautic.DEBUG: CAMPAIGN: Triggering kickoff events  
[2021-03-25 07:00:11] mautic.DEBUG: CAMPAIGN: Processing the following events: 1  
[2021-03-25 07:00:11] mautic.DEBUG: CAMPAIGN: No events to process  
[2021-03-25 07:00:11] mautic.DEBUG: CAMPAIGN: Triggering scheduled events  
[2021-03-25 07:00:11] mautic.DEBUG: CAMPAIGN: 1 events scheduled to execute.  
[2021-03-25 07:00:11] mautic.DEBUG: CAMPAIGN: Comparing calculated executed time for event ID 2 and contact ID 1 with 2021-03-22 17:16:47 UTC  
[2021-03-25 07:00:11] mautic.DEBUG: CAMPAIGN: Scheduling event ID 2 for contact ID 1 based on hour of 07:00 UTC  
[2021-03-25 07:00:11] mautic.DEBUG: CAMPAIGN: Scheduling event ID 2 for contact ID 1 based on DOW restrictions of 4  
[2021-03-25 07:00:11] mautic.DEBUG: CAMPAIGN: Log ID #2 to be executed on 2021-03-25 07:00:00 UTC compared to 2021-03-25 07:00:11 UTC  
[2021-03-25 07:00:11] mautic.DEBUG: CAMPAIGN: Executing email.send ID 2  
[2021-03-25 07:00:12] mautic.DEBUG: CAMPAIGN: Evaluating 0 conditions for action ID 2  
[2021-03-25 07:00:12] mautic.DEBUG: CAMPAIGN: Executing 1 actions under action ID 2  
[2021-03-25 07:00:12] mautic.DEBUG: CAMPAIGN: (3) Adding interval of 1d to 2021-03-25 07:00:11 UTC  
[2021-03-25 07:00:12] mautic.DEBUG: CAMPAIGN: (3) 2021-03-26 07:00:11 UTC is later than 2021-03-25 07:00:11 UTC and thus returning 2021-03-26 07:00:11 UTC  
[2021-03-25 07:00:12] mautic.DEBUG: CAMPAIGN: Event ID# 3 to be executed on 2021-03-26 07:00:11 UTC  
[2021-03-25 07:00:12] mautic.DEBUG: CAMPAIGN: Action ID# 3 for contact ID# 1 has timing that is not appropriate and thus scheduled for 2021-03-26 07:03:00 UTC  
[2021-03-25 07:00:12] mautic.DEBUG: CAMPAIGN: Triggering inaction events  
[2021-03-25 07:00:12] mautic.DEBUG: CAMPAIGN: No events to process  
```

### 25/03/2021: subscribed again the same contact to the campaign
* First email is sent
* Database dump:
![snap3](https://user-images.githubusercontent.com/101675/112033403-a99f7e00-8b3d-11eb-9a5d-8d2a4aabfcf3.png)
* Log:
```
[2021-03-25 07:01:13] mautic.DEBUG: CONTACT: Tracking session for contact ID#  through POST /mautic/form/submit?formId=1  
[2021-03-25 07:01:13] mautic.DEBUG: CONTACT: Tracking session for contact ID#  through POST /mautic/form/submit?formId=1  
[2021-03-25 07:01:13] mautic.DEBUG: FORM: Not in kiosk mode so using current contact ID #  
[2021-03-25 07:01:13] mautic.DEBUG: FORM: Unique fields submitted include test@mascellani.eu  
[2021-03-25 07:01:13] mautic.DEBUG: 1 found based on unique identifiers  
[2021-03-25 07:01:13] mautic.DEBUG: FORM: Testing contact ID# 1 for conflicts  
[2021-03-25 07:01:13] mautic.DEBUG: FORM: Potential conflicts  =   
[2021-03-25 07:01:13] mautic.DEBUG: FORM: In kiosk mode so not merging  
[2021-03-25 07:01:13] mautic.DEBUG: FORM: Potential conflicts 0 = email  
[2021-03-25 07:01:13] mautic.DEBUG: FORM: Current unique contact fields email = test@mascellani.eu  
[2021-03-25 07:01:13] mautic.DEBUG: FORM: Submitted unique contact fields email = test@mascellani.eu  
[2021-03-25 07:01:13] mautic.DEBUG: FORM: Associating 127.0.0.1 to contact  
[2021-03-25 07:01:13] mautic.DEBUG: CONTACT: 1 set as current lead.  
[2021-03-25 07:01:14] mautic.DEBUG: LEAD: Tracking ID for this device is ij2t0yb92z6o2i2mnxdl5b5  
[2021-03-25 07:01:14] mautic.DEBUG: CONTACT: Tracking code changed from  for contact ID#  to ij2t0yb92z6o2i2mnxdl5b5 for contact ID# 1  
[2021-03-25 07:01:14] mautic.DEBUG: CAMPAIGN: Membership for contact ID 1 in campaign ID 1 was updated to be included.  
[2021-03-25 07:01:14] mautic.DEBUG: CONTACT: Tracking session for contact ID# 1 through POST /mautic/form/submit?formId=1  
[2021-03-25 07:01:14] mautic.DEBUG: CAMPAIGN: Campaign triggered for event type form.submit(form / 1)  
[2021-03-25 07:01:14] mautic.DEBUG: CONTACT: Tracking session for contact ID# 1 through POST /mautic/form/submit?formId=1  
[2021-03-25 07:01:14] mautic.DEBUG: CAMPAIGN: Current contact ID# 1  
[2021-03-25 07:01:14] mautic.DEBUG: CAMPAIGN: Contact does not have any applicable form.submit associations.  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Triggering kickoff events  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Processing the following events: 1  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Processing the following contacts: 1  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: (1) Executing immediately  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Event ID# 1 to be executed on 2021-03-25 07:01:17 UTC compared to 2021-03-25 07:01:17 UTC  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: (1) Executing immediately  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Event ID# 1 to be executed on 2021-03-25 07:01:17 UTC  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Executing email.send ID 1  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Evaluating 0 conditions for action ID 1  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Executing 1 actions under action ID 1  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: (2) Adding interval of 1d to 2021-03-25 07:01:17 UTC  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: (2) 2021-03-26 07:01:17 UTC is later than 2021-03-25 07:01:17 UTC and thus returning 2021-03-26 07:01:17 UTC  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Event ID# 2 to be executed on 2021-03-26 07:01:17 UTC  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Comparing calculated executed time for event ID 2 and contact ID 1 with 2021-03-25 07:01:17 UTC  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Scheduling event ID 2 for contact ID 1 based on hour of 07:00 UTC  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Scheduling event ID 2 for contact ID 1 based on DOW restrictions of 4  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Action ID# 2 for contact ID# 1 has timing that is not appropriate and thus scheduled for 2021-04-01 07:04:00 UTC  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Fetching the next batch of kickoff contacts starting with contact ID 2  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: No more contacts to process  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Triggering scheduled events  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: 0 events scheduled to execute.  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: No events to process  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: Triggering inaction events  
[2021-03-25 07:01:17] mautic.DEBUG: CAMPAIGN: No events to process  
```

### 26/03/2021 (Friday)
* Third email is sent
* Database dump:
![snap4](https://user-images.githubusercontent.com/101675/112033560-d358a500-8b3d-11eb-8ded-b46e77c1a19a.png)
* Log:
```
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: Triggering kickoff events  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: Processing the following events: 1  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: No events to process  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: Triggering scheduled events  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: 1 events scheduled to execute.  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: (3) Adding interval of 1d to 2021-03-25 07:00:12 UTC  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: (3) 2021-03-26 07:00:12 UTC is earlier than 2021-03-26 07:01:14 UTC and thus returning 2021-03-26 07:01:14 UTC  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: Log ID #3 to be executed on 2021-03-26 07:01:14 UTC compared to 2021-03-26 07:01:14 UTC  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: Executing email.send ID 3  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: Evaluating 0 conditions for action ID 3  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: Executing 1 actions under action ID 3  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: (4) Adding interval of 1d to 2021-03-26 07:01:14 UTC  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: (4) 2021-03-27 07:01:14 UTC is later than 2021-03-26 07:01:14 UTC and thus returning 2021-03-27 07:01:14 UTC  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: Event ID# 4 to be executed on 2021-03-27 07:01:14 UTC  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: Action ID# 4 for contact ID# 1 has timing that is not appropriate and thus scheduled for 2021-03-27 07:03:01 UTC  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: Triggering inaction events  
[2021-03-26 07:01:14] mautic.DEBUG: CAMPAIGN: No events to process  
```

### 27/03/2021 (Saturday)
* Fourth email is sent
* Database dump:
![snap5](https://user-images.githubusercontent.com/101675/112033642-ecf9ec80-8b3d-11eb-8118-3a3b7c7decaa.png)
* Log:
```
[2021-03-27 07:02:12] mautic.DEBUG: CAMPAIGN: Triggering kickoff events  
[2021-03-27 07:02:12] mautic.DEBUG: CAMPAIGN: Processing the following events: 1  
[2021-03-27 07:02:12] mautic.DEBUG: CAMPAIGN: No events to process  
[2021-03-27 07:02:12] mautic.DEBUG: CAMPAIGN: Triggering scheduled events  
[2021-03-27 07:02:12] mautic.DEBUG: CAMPAIGN: 1 events scheduled to execute.  
[2021-03-27 07:02:12] mautic.DEBUG: CAMPAIGN: (4) Adding interval of 1d to 2021-03-26 07:01:14 UTC  
[2021-03-27 07:02:12] mautic.DEBUG: CAMPAIGN: (4) 2021-03-27 07:01:14 UTC is earlier than 2021-03-27 07:02:12 UTC and thus returning 2021-03-27 07:02:12 UTC  
[2021-03-27 07:02:12] mautic.DEBUG: CAMPAIGN: Log ID #6 to be executed on 2021-03-27 07:02:12 UTC compared to 2021-03-27 07:02:12 UTC  
[2021-03-27 07:02:12] mautic.DEBUG: CAMPAIGN: Executing email.send ID 4  
[2021-03-27 07:02:12] mautic.DEBUG: CAMPAIGN: Evaluating 0 conditions for action ID 4  
[2021-03-27 07:02:12] mautic.DEBUG: CAMPAIGN: Executing 0 actions under action ID 4  
[2021-03-27 07:02:12] mautic.DEBUG: CAMPAIGN: Triggering inaction events  
[2021-03-27 07:02:12] mautic.DEBUG: CAMPAIGN: No events to process  
```

### 01/04/2021 (Thursday)
* Second email is sent
* Database dump:
![snap6](https://user-images.githubusercontent.com/101675/112033723-069b3400-8b3e-11eb-9019-9438b1239db8.png)
* Log:
```
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: Triggering kickoff events  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: Processing the following events: 1  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: No events to process  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: Triggering scheduled events  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: 1 events scheduled to execute.  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: Comparing calculated executed time for event ID 2 and contact ID 1 with 2021-03-25 07:01:17 UTC  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: Scheduling event ID 2 for contact ID 1 based on hour of 07:00 UTC  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: Scheduling event ID 2 for contact ID 1 based on DOW restrictions of 4  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: Log ID #5 to be executed on 2021-04-01 07:00:00 UTC compared to 2021-04-01 07:00:14 UTC  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: Executing email.send ID 2  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: Evaluating 0 conditions for action ID 2  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: Executing 1 actions under action ID 2  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: (3) Adding interval of 1d to 2021-04-01 07:00:14 UTC  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: (3) 2021-04-02 07:00:14 UTC is later than 2021-04-01 07:00:14 UTC and thus returning 2021-04-02 07:00:14 UTC  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: Event ID# 3 to be executed on 2021-04-02 07:00:14 UTC  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: Action ID# 3 for contact ID# 1 has timing that is not appropriate and thus scheduled for 2021-04-02 07:04:00 UTC  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: Triggering inaction events  
[2021-04-01 07:00:14] mautic.DEBUG: CAMPAIGN: No events to process  
```

### 02/04/2021 (Friday)
* Third email is sent
* Trigger script crashes and doesn't schedule Saturday's email
* Database dump:
![snap7](https://user-images.githubusercontent.com/101675/112033876-2894b680-8b3e-11eb-890c-f627ee86843d.png)
* Log:
```
[2021-04-02 07:01:13] mautic.DEBUG: CAMPAIGN: Triggering kickoff events  
[2021-04-02 07:01:13] mautic.DEBUG: CAMPAIGN: Processing the following events: 1  
[2021-04-02 07:01:13] mautic.DEBUG: CAMPAIGN: No events to process  
[2021-04-02 07:01:13] mautic.DEBUG: CAMPAIGN: Triggering scheduled events  
[2021-04-02 07:01:13] mautic.DEBUG: CAMPAIGN: 1 events scheduled to execute.  
[2021-04-02 07:01:13] mautic.DEBUG: CAMPAIGN: (3) Adding interval of 1d to 2021-04-01 07:00:14 UTC  
[2021-04-02 07:01:13] mautic.DEBUG: CAMPAIGN: (3) 2021-04-02 07:00:14 UTC is earlier than 2021-04-02 07:01:13 UTC and thus returning 2021-04-02 07:01:13 UTC  
[2021-04-02 07:01:13] mautic.DEBUG: CAMPAIGN: Log ID #7 to be executed on 2021-04-02 07:01:13 UTC compared to 2021-04-02 07:01:13 UTC  
[2021-04-02 07:01:13] mautic.DEBUG: CAMPAIGN: Executing email.send ID 3  
[2021-04-02 07:01:13] mautic.DEBUG: CAMPAIGN: Evaluating 0 conditions for action ID 3  
[2021-04-02 07:01:13] mautic.DEBUG: CAMPAIGN: Executing 1 actions under action ID 3  
[2021-04-02 07:01:13] mautic.DEBUG: CAMPAIGN: (4) Adding interval of 1d to 2021-04-02 07:01:13 UTC  
[2021-04-02 07:01:13] mautic.DEBUG: CAMPAIGN: (4) 2021-04-03 07:01:13 UTC is later than 2021-04-02 07:01:13 UTC and thus returning 2021-04-03 07:01:13 UTC  
[2021-04-02 07:01:13] mautic.DEBUG: CAMPAIGN: Event ID# 4 to be executed on 2021-04-03 07:01:13 UTC  
[2021-04-02 07:01:13] mautic.DEBUG: CAMPAIGN: Action ID# 4 for contact ID# 1 has timing that is not appropriate and thus scheduled for 2021-04-03 07:04:01 UTC  
[2021-04-02 07:01:13] mautic.ERROR: CAMPAIGN: An exception occurred while executing 'INSERT INTO campaign_lead_event_log (rotation, date_triggered, is_scheduled, trigger_date, system_triggered, metadata, channel, channel_id, non_action_path_taken, event_id, lead_id, campaign_id, ip_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [2, "2021-04-02 07:01:13", 1, "2021-04-03 07:01:13", 1, "a:0:{}", null, null, 0, 4, "1", 1, null]:

SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '4-1-2' for key 'campaign_rotation'  
```

The behaviour is clearly not correct: email are sent in an erratic order, eventually the campaign stalls without reaching the end and of course the databse should not give the integrity constraing violation error.

**Comment by a-sandris** [comment]
Looks like we have the same issue at the moment.

**Comment by giomasce** [comment]
I pushed a possible fix [here](https://github.com/giomasce/mautic/commit/e64a1d469cd1cbfa752511750a0c7d0eed2d4bae). It's probably quite bad code style because I have very little experience with PHP, but it might get the job done.

**Comment by RCheesley** [comment]
@giomasce could you please make a PR from your changes and link it to this issue by saying 'fixes #9797' in the description? That way we can review the code and discuss it in the PR.

Here's some docs on making a PR in case it is helpful: https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests

**Comment by a-sandris** [comment]
We have applied a `possible fix` and so far looks like the error is gone. @giomasce please create a PR so that we have a chance to get that in the core.

**Comment by giomasce** [comment]
[There it is!](https://github.com/mautic/mautic/pull/9829)

**Comment by kuzmany** [comment]
@giomasce cab you show your campaign from comment https://github.com/mautic/mautic/issues/9797#issuecomment-804262839 

That actions are chained or not?

**Comment by giomasce** [comment]
Hi @kuzmany, does this snapshot answer your question? A more detailed description of the campaign is in the bug description.

![image](https://user-images.githubusercontent.com/101675/112531047-6a746580-8da7-11eb-9535-3478baf37040.png)

**Comment by kuzmany** [comment]
@giomasce yes, that's chained  type

I suppose to this kind of workflow should works

![image](https://user-images.githubusercontent.com/462477/112595679-0c796980-8e0b-11eb-97dc-711abeaa12fb.png)

**Comment by kuzmany** [comment]
@giomasce Did you setup also something  from relative time settings? (hours, days...)
![image](https://user-images.githubusercontent.com/462477/112596591-4eef7600-8e0c-11eb-8dff-9d60aa1bf2b2.png)

**Comment by giomasce** [comment]
This is for the second event: 
![image](https://user-images.githubusercontent.com/101675/112603953-f2915400-8e15-11eb-9653-060aee546ea4.png)

Third event:
![image](https://user-images.githubusercontent.com/101675/112604029-050b8d80-8e16-11eb-8e23-ad1509ac6730.png)

Fourth event:
![image](https://user-images.githubusercontent.com/101675/112604078-105eb900-8e16-11eb-86f2-f06b930aeb25.png)

**Comment by giomasce** [comment]
> I suppose to this kind of workflow should works

I don't understand, are you asking me to test? I can test that, but whatever my test gives, the bug I found is still there. I believe Mautic should never incur in an integrity database error and campaigns stalling, even with my graph.

**Comment by kuzmany** [comment]
I've just tried understand what is issue. Now I think understand it.
You expect chained actions are scheduled based on action above. 

Chained actions was added by this PR

https://github.com/mautic/mautic/pull/6187

Need investigate more later

**Comment by giomasce** [comment]
Actually, I don't think that the problem is with chained actions. They seem to work well. The problem happens when a contact, which is *already* in the campaign, subscribes again. What I expect is that the campaign restarts from the beginning, as the contact were subscribing for the first time, and therefore the currently scheduled events in the campaign are unscheduled. But this does not happen, and the contact has then scheduled events both from the first and second campaign rotation (see my database dump above).

**Comment by giomasce** [comment]
> This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

This issue is perfectly valid, and not stale at all. Please, review the PR instead of hiding the bug.

**Comment by RCheesley** [comment]
Thanks for keeping the issue active, it looks like PHPSTAN was failing on your PR which is probably why it hasn't been reviewed yet. I've just pulled in the latest updates from the Features branch.

If you are able to check whether this is still an issue in Mautic 4 (features branch) that would be a great help.

In the meantime, perhaps you could help us by testing some of the 210 open pull requests? That would also be extremely helpful for developers who are also waiting for their PRs to be tested.

**Comment by giomasce** [comment]
I tried to reproduce, but after setting up again the test environment I hit something that seems to be https://github.com/mautic/mautic/issues/9951, and could not proceed. Unfortunately I don't have a lot of time, so for the moment this has to wait on my side. Sorry!

**Comment by giomasce** [comment]
Not stale.

**Comment by dongilbert** [comment]
Thanks for the detail here @giomasce - I've approved #9829 to fix this

**Comment by giomasce** [comment]
Not stale, I'll keep working on it as soon as I have time.

**Comment by giomasce** [comment]
Not stale.

**Comment by RCheesley** [comment]
Thanks for letting us know @giomasce - hopefully we can get the tests written and get it merged soon! The PR will need a rebase though, bug fixes will be released from the 4.4 branch at this time.

**Comment by giomasce** [comment]
The issue is still applicable. I don't understand what's the point of closing bug reports that still apply, but apparently there is nothing I can do. If I don't reply to stalebot's pings, then the issue gets closed. If I reply, it gets closed anyway. Very puzzling.
